### PR TITLE
Change our yield implementation to allow caller frames to be efficiently accessed by blocks when required.

### PR DESCRIPTION
**Pull request by aardvark179**

### CODE REVIEW

File: src/main/ruby/core/object_space.rb [hunk 1]
```diff
@@ -28,4 +28,15 @@ module ObjectSpace
   def self.garbage_collect
     GC.start
   end
+
+  def self.each_object( of_class=nil )
```
Comment (chrisseaton) [inline src/main/ruby/core/object_space.rb]: Strange spaces.

File: src/main/ruby/core/object_space.rb [hunk 2]
```diff
@@ -28,4 +28,15 @@ module ObjectSpace
...
+    objects = Truffle::ObjectSpaceOperations.all_objects(of_class)
+
+    return to_enum(:each_object, of_class) { objects.size } unless block_given?
+    
+    objects.each do |o|
+      yield o
+    end
+    objects.size
```
Comment (chrisseaton) [inline src/main/ruby/core/object_space.rb]: Isn't there some utility method that does all these things?

File: src/main/java/org/truffleruby/language/yield/CallBlockNode.java [hunk 3]
```diff
@@ -98,5 +189,4 @@ protected DirectCallNode createBlockCallNode(CallTarget callTarget) {
     protected int getCacheLimit() {
         return getContext().getOptions().YIELD_CACHE;
     }
-
```
Comment (chrisseaton) [inline src/main/java/org/truffleruby/language/yield/CallBlockNode.java]: No need to delete this line.

File: src/main/java/org/truffleruby/core/objectspace/TruffleObjectSpaceNodes.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
+package org.truffleruby.core.objectspace;
```
Comment (chrisseaton) [inline src/main/java/org/truffleruby/core/objectspace/TruffleObjectSpaceNodes.java]: Copyright header.

File: src/main/java/org/truffleruby/core/kernel/TraceManager.java [hunk 5]
```diff
@@ -169,12 +169,12 @@ protected void onEnter(VirtualFrame frame) {
...
             try {
-                yield(traceFunc, event,
+                yield(frame, traceFunc,
+                        event,
                         getFile(file),
                         line,
                         context.getSymbolTable().getSymbol(RubyArguments.getMethod(frame).getName()),
-                        BindingNodes.createBinding(context, frame.materialize()),
-                        getLogicalClass(RubyArguments.getSelf(frame)));
+                        BindingNodes.createBinding(context, frame.materialize()), getLogicalClass(RubyArguments.getSelf(frame)));
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/kernel/TraceManager.java]: One argument per line here.

File: src/main/java/org/truffleruby/core/objectspace/ObjectSpaceNodes.java [hunk 6]
```diff
@@ -98,38 +102,38 @@ protected boolean isFloatID(DynamicObject id) {
...
+            for (DynamicObject object : objects) {
                 if (!isHidden(object)) {
-                    yield(block, object);
-                    count++;
+                    objectList.add(object);
                 }
             }
 
-            return count;
+            return createArray(objectList.toArray(), objectList.size());
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/objectspace/ObjectSpaceNodes.java]: This is very significant overhead: it's a second list of all DynamicObject in the heap!

File: src/main/java/org/truffleruby/core/thread/ThreadNodes.java [hunk 7]
```diff
@@ -503,12 +505,14 @@ public DynamicObject wakeup(DynamicObject rubyThread,
     @CoreMethod(names = "unblock", required = 2)
     public abstract static class UnblockNode extends YieldingCoreMethodNode {
 
+        static final VirtualFrame NEW_FRAME = Truffle.getRuntime().createVirtualFrame(new Object[0], new FrameDescriptor());
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/thread/ThreadNodes.java]: That's a ugly hack.

File: src/main/ruby/core/configuration.rb [hunk 8]
```diff
@@ -37,7 +37,14 @@ def get_variable(name)
     end
 
     def section(section, &block)
-      Truffle.invoke_primitive :vm_get_config_section, section, block
+      a_config = Truffle.invoke_primitive :vm_get_config_section, section
+      limit = a_config.size
+      # We don't have all the standard enumerable methods at this point, so can't use each_slice.
+      index = 0
+      while (index < limit) do
```
Comment (eregon) [inline src/main/ruby/core/configuration.rb]: No `do` and no `()`.

File: src/main/ruby/core/encoding.rb [hunk 9]
```diff
@@ -74,7 +74,7 @@ def build_encoding_map
 
     def build_transcoding_map
       map = {}
-      Encoding::Converter.each_transcoder do |source, destinations|
+      all = Encoding::Converter.all_transcoders.each_slice(2) do |source, destinations|
```
Comment (eregon) [inline src/main/ruby/core/encoding.rb]: This sounds like it's going to be fairly costly in startup.

File: src/main/ruby/core/object_space.rb [hunk 10]
```diff
@@ -28,4 +28,15 @@ module ObjectSpace
   def self.garbage_collect
     GC.start
   end
+
+  def self.each_object( of_class = nil )
```
Comment (eregon) [inline src/main/ruby/core/object_space.rb]: No extra spaces.

File: src/main/java/org/truffleruby/core/encoding/TruffleEncodingNodes.java [hunk 11]
```diff
@@ -33,6 +41,31 @@
 @CoreClass("Truffle::EncodingOperations")
 public abstract class TruffleEncodingNodes {
 
+    @NonStandard
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/encoding/TruffleEncodingNodes.java]: No need to annotate with `@NonStandard` anymore if it's under `Truffle::`.

**Comment by pitr-ch** [comment]
I have no issue with passing a frame around for calling a block. Looks good to me.

**Comment by eregon** [comment]
@aardvark179 This should no longer be needed thanks to #930, right?
We can of course revisit if there are other needs for fast block caller frames.

**Comment by aardvark179** [comment]
Closing this as it all got done a different way.

**Pull request closed without merging**